platform/logger.h

```diff
@@ -0,0 +1,150 @@
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t len##counter = MBED_STRLEN(fmt); \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t c##counter = id; \
+                                                       MBED_LOG_ID_4(args, id, ##__VA_ARGS__); \
+                                                      })
+#else
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)    MBED_LOG_ID_4(args, id, ##__VA_ARGS__);
+#endif
+
+#define MBED_LOG_ID_2(counter, id, ...)               MBED_LOG_ID_3(counter, id, MBED_COUNT_VA_ARGS(__VA_ARGS__), ##__VA_ARGS__)
+#define MBED_LOG_ID_1(mod, fmt, ll, f, l, c, ...)     MBED_LOG_ID_2(c, TRACE_ID_(mod,c,l), "[" mod "][" f "][" MBED_STRINGIFY(l) "]: " fmt, ##__VA_ARGS__)
+
+// Macros to log string data
+#define MBED_LOG_STR(...)                             log_buffer_string_data(__VA_ARGS__)
+#define MBED_LOG_STR_1(mod, fmt, ll, ...)             MBED_LOG_STR("[%-4.4s][%-4.4s]: " fmt, ll, mod, ##__VA_ARGS__)
+
+// Assert string
+#define MBED_LOG_ASSERT(...)                          log_assert(__VA_ARGS__)
+#define MBED_LOG_ASSERT_1(mod, fmt, ll, f, l, ...)    MBED_LOG_ASSERT("[%-4.4s][%-4.4s][%-15s][%5d]: " fmt, ll, mod, f, l, ##__VA_ARGS__)
+
+#define MBED_LOG_VSTR_2(fmt1, ap)                     log_buffer_string_vdata(fmt1, ap)
+#define MBED_LOG_VSTR_1(ll, mod, fmt, ap)             MBED_LOG_VSTR_2("[" ll "][" mod "]: " fmt, ap)
+
+void log_buffer_id_data(uint32_t argCount, ...);
+void log_buffer_string_data(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+void log_buffer_string_vdata(const char *format, va_list args);
+void log_buffer_string_vdata_critical(const char *format, va_list args, uint8_t lossy);
+void log_assert(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+
+// Functions added to test the logging feature
+void log_reset(void);
+uint32_t log_get_bytes_lost(void);
+void log_buffer_data(char *str);
+void log_disable_time_capture(void);
+void log_enable_time_capture(void);
+
+#if defined(NDEBUG) || defined(MBED_ID_BASED_TRACING)
+#define mbed_log_array
+#define mbed_log_ipv6
+#define mbed_log_ipv6_prefix
+#define mbed_log_helper_lock
+#define mbed_log_helper_unlock
+#define mbed_log_helper_unlock_all
+#define mbed_log_valid_helper_data
+#define mbed_log_get_helper_data
+#else
+// Helper functions : Note helper functions are not ISR safe.
+char* mbed_log_array(const uint8_t* buf, uint32_t len);
+char* mbed_log_ipv6(const uint8_t* addr_ptr);
+char* mbed_log_ipv6_prefix(const uint8_t* prefix, uint32_t prefix_len);
+
+void mbed_log_helper_lock(void);
+void mbed_log_helper_unlock(void);
+void mbed_log_helper_unlock_all(void);
+int mbed_log_valid_helper_data(void);
+char* mbed_log_get_helper_data(void);
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif  // LOGGER_INTERNAL_H
```

Comment:
file is named logger but this says internal? Is this internal header file?

Reply:
Yes will rename it to logger_internal.h